.github/workflows/ci.yml

```diff
@@ -24,69 +22,50 @@ jobs:
           profile: minimal
           toolchain: ${{ matrix.rust }}
           override: true
-          components: rustfmt, clippy
-      - uses: actions-rs/cargo@v1
-        with:
-          command: test
-  
-  coverage_and_publish:
-    runs-on: ubuntu-latest
-    steps:
-      - uses: actions/checkout@v2
-      - uses: actions-rs/toolchain@v1
-        with:
-          profile: minimal
-          toolchain: stable
-          override: true
-      - name: Generate coverage info
-        uses: actions-rs/tarpaulin@v0.1
-        with:
-          version: '0.18.0'
-          args: '-- --test-threads 1'
-          out-type: 'Lcov'
-      - name: Upload to coveralls.io
-        uses: coverallsapp/github-action@master
-        with:
-          github-token: ${{secrets.GITHUB_TOKEN}}
-          path-to-lcov: './lcov.info'
-      - name: Publish to crates.io
-        if: github.event_name == 'release'
+      - name: Test
         uses: actions-rs/cargo@v1
         with:
-          command: publish
-          args: --token ${{ secrets.CRATES_TOKEN }}
+          command: test
 
-  fmt:
-    name: Rustfmt
+  lint:
     runs-on: ubuntu-latest
-    if: github.event_name != 'release'
     steps:
       - uses: actions/checkout@v2
       - uses: actions-rs/toolchain@v1
         with:
           profile: minimal
           toolchain: stable
           override: true
-      - run: rustup component add rustfmt
-      - uses: actions-rs/cargo@v1
+          components: rustfmt, clippy
+      - name: Formatting (fmt)
+        uses: actions-rs/cargo@v1
         with:
           command: fmt
           args: --all -- --check
+      - name: Clippy
+        uses: actions-rs/cargo@v1
+        with:
+          command: clippy
+          args: -- -D warnings
```

Comment:
It's better to run clippy after building.
* Clippy is faster because it can reuse build artifacts
* It actually makes sense, because clippy thoulg also be run for each supported `--target` like apple, linux or windows.

Reply:
@maxammann i agree in principle with clippy. The issue here is that there are two independent VMs - the `test` and `lint`. `Test` builds on multiple rust versions - so there is no point to run clippy on all of them. Also, it does not install fmt and clippy.  `Lint` runs only on one Rust version, but with extra features (fmt, clippy). So it kinda makes sense to keep them separate.  One alternative of course would be to make a conditional fmt and clippy steps that only run if the `matrix.rust == "stable"`

Reply:
Hm in that case clippy should also run for each toolchain im the matrix.

It would be that clippy in nightly does not accept the current code. That would be an issue and in fact that would indicate that the crate is not compatible with e.g. nightly.

I would estimate that clippy takes max. 30 seconds after cargo test has run.

Reply:
I found clippy to only be useful in the stable tbh. Afterall, its a recommendation system, so I would rather not run clippy beyond just one verison. In theory I can even imagine a case where clippy in one version has an opposing suggestion to another version.

Reply:
I think we can polish this a bit later too - my first goal is to get CI to run. Currently its off

Reply:
Sure, yeah that case with opposing suggestions would be weird I agree :D